### PR TITLE
ERC721 Weighted Pairs [erc721-pair-weights]

### PR DESCRIPTION
**Pull request by arpitkarnatak**
# ERC721 Weighted Pairs

This strategy determines the voting powers for a pair of ERC721 NFT registries. This strategy supports 2 different NFT collections and it provides different voting powers to the holders of these collections, and also supports bonus powers for holders which have items from both the collections.

```json
{
    "symbol": "OCMONK",
    "registries": [
      "0x960b7a6bcd451c9968473f7bbfd9be826efd549a",
      "0x86cc280d0bac0bd4ea38ba7d31e895aa20cceb4b"
    ],
    "weights": [
      2,
      3
    ],
    "pairWeight": 4 
 }
```

In `example.json`, address `0xE052113bd7D7700d623414a0a4585BCaE754E9d5` has 31 items of `collection1` NFT and 35 items of `collection2` NFT. 

As we can see, there are 31 pairs, 0 items exclusively for collection1 and 4 items exclusively for collection2. So, the voting power of the address will be 

```
(31 * pairWeight) + (collection1 * weights[0]) + (collection2 * weights[1])`
= 31*4 + 0*2 + 4*3
= 136
```
which can be seen in the snapshot:

```
    [
      {
        '0x5F9345FdAd91Bf9757445ADc82e477B33FD3349c': 6,
        '0xE052113bd7D7700d623414a0a4585BCaE754E9d5': 136
      }
    ]
```

### CODE REVIEW

**Review by rudi-cilibrasi — APPROVED** [review]

**Comment by huuep** [comment]
fyi, the project that plans to use this strategy is OnChainMonkey:
https://onchainmonkey.com/dao
https://twitter.com/onchainmonkey
Is there anything else you need from us to get this strategy approved? Thanks!

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/strategies/erc721-pair-weights/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    "pairWeight": 4 
+ }
+```
+
+In `example.json`, address `0xE052113bd7D7700d623414a0a4585BCaE754E9d5` has 31 items of `collection1` NFT and 35 items of `collection2` NFT. 
+
+As we can see, there are 31 pairs, 0 items exclusively for collection1 and 4 items exclusively for collection2. So, the voting power of the address will be 
+
+```
+(31 * pairWeight) + (collection1 * weights[0]) + (collection2 * weights[1])`
```
Comment (ChaituVR) [inline src/strategies/erc721-pair-weights/README.md:25]: There is a bit confusion with strategy name and calculation here 😄 main aim is to add weight for two different collections, but here main priority is given to pair weight than collect weight

Comment (arpitkarnatak) [reply src/strategies/erc721-pair-weights/README.md:25]: Yes, the idea was that someone holding from both the collections should have more priority than the addresses that hold items from only one of the two collections. If someone wants no additional priority for the addresses holding pairs, we can assign the values as `pairWeight*2 = weights[0] = weights[1]`

Comment (rudi-cilibrasi) [reply src/strategies/erc721-pair-weights/README.md:25]: it's supposed to give a "bonus" vote(s) if you hold BOTH a genesis and a karma NFT. and more pairs is more bonus. depends on parameters. we can even set weights to penalize pair holders. but we don't expect to use that in our case.

File: src/strategies/erc721-pair-weights/index.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  network,
+  provider,
+  addresses,
+  options,
+  snapshot
+) {
+  const blockTag = typeof snapshot === 'number' ? snapshot : 'latest';
+
+  const calls: any[] = [];
+  options.registries.forEach((registry) => {
```
Comment (ChaituVR) [inline src/strategies/erc721-pair-weights/index.ts:22]: Limiting it to 2 registries for now (as we are using only 2 balances below code), so others won't be able to pass too many registries to break something, maybe in the future, we can increase :)
```suggestion
  options.registries.slice(0,2).forEach((registry) => {
```

**Comment by rudi-cilibrasi** [comment]
please let us know if you still want us to change anything @ChaituVR and thanks for the fast review and resolutions.

**Review by ChaituVR — APPROVED** [review]

**Comment by huuep** [comment]
Thanks @ChaituVR!

**Pull request merged**